File: CHUOND_4/Activity_Result_Kotlin_Ex3/app/src/main/java/com/example/ex3/MainActivity.java
package com.example.ex3;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.provider.ContactsContract;
import android.widget.Toast;

import androidx.activity.EdgeToEdge;
import androidx.activity.result.ActivityResult;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

import com.example.ex3.databinding.ActivityMainBinding;

public class MainActivity extends AppCompatActivity {
    private ActivityMainBinding mBinding;
    private final ActivityResultLauncher<Intent> mLauncher = registerForActivityResult(
            new ActivityResultContracts.StartActivityForResult(),
            this::handleResult
    );

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        mBinding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(mBinding.getRoot());
        ViewCompat.setOnApplyWindowInsetsListener(mBinding.main, (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });
        setupViews();
    }

    private void handleResult(ActivityResult result) {
        Uri uri = result.getData() != null ? result.getData().getData() : null;
        if (uri != null && result.getResultCode() == RESULT_OK){
            Intent intent = new Intent(Intent.ACTION_VIEW, uri);
            startActivity(intent);
        } else {
            Toast.makeText(
                    this,
                    getString(R.string.message_no_data),
                    Toast.LENGTH_LONG
            ).show();
        }
    }

    private void setupViews(){
        mBinding.btnStartActivity.setOnClickListener(v -> {
            Intent targetIntent = new Intent(Intent.ACTION_PICK, ContactsContract.Contacts.CONTENT_URI);
            mLauncher.launch(targetIntent);
        });
    }
}

File: CHUOND_4/Ex1_ViewModel_Kotlin/app/src/main/java/com/example/ex1/MainActivity.kt
package com.example.ex1

import android.os.Bundle
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.lifecycle.ViewModel
import com.example.ex1.databinding.ActivityMainBinding
import kotlin.time.Duration

class MainActivity : AppCompatActivity() {
    private val viewModel: FullNameViewModel by viewModels()
    private lateinit var binding: ActivityMainBinding
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }
        setupViewModel()
        setupView()
    }

    private fun setupViewModel() {
        viewModel.fullName.observe(this){
            binding.tvFullName.text = getString(R.string.full_name, it)
        }
    }

    private fun setupView() {
        binding.btnSubmit.setOnClickListener {
            viewModel.setCurrentFullName(binding.etFullName.text.toString())
            Toast.makeText(this, binding.tvFullName.text.toString(), Toast.LENGTH_LONG).show()
        }
    }
}